Secrets rotations Runbooks

rogeruiz opened the pull request:
Pairing with @wjwoodson ~~and @commit-dkp~~ on the ~~initial~~ work done here. @18F/cloud-gov-ops, please comment or add to this PR as needed.

rogeruiz committed:
Move the rotating-iam-users under runbook/

rogeruiz committed:
Initial draft for runbooks/

jmcarp commented on content/docs/ops/runbook/rotating-secrets.md:
```diff
@@ -0,0 +1,44 @@
+- Rotate these specific IAM Users first
+    - `bosh_user`
+    - `concourse_user`
+    - `cf_user`
```
We dropped `cf_user` in favor of instance profiles recently, and for some reason we've never actually provisioned the read-only admin user, so there aren't currently any credentials rotate there. Maybe it's time to either drop that user or add it to a stack--wdyt @LinuxBozo ?

rogeruiz replied:
yeah good call on both of these. I wonder if the runbook is the best place for the actual names? or if it's just better to say something like. **Look through `cg-provision`, and in AWS, and make sure any IAM users with Policies attached to them that can modify contents of buckets or the infrastructure need to be rotated first?

jmcarp replied:
By the end of the quarter, we're supposed to get rid of all IAM users, so this shouldn't matter for long. But for now, 👍 to "rotate all the users" or similar.

rogeruiz committed:
More progress on CF/Diego docs

rogeruiz committed:
More drafts for general rotation work

rogeruiz committed:
Finish up Cloud Foundry and Diego docs

rogeruiz committed:
Small updates

rogeruiz committed:
Update secrets documentation around multiple files

rogeruiz committed:
Still drafting secrets management docs;

This is to help match what our secrets look like now after splitting
them up into `*.main.yml` and `*.external.yml` files.

rogeruiz committed:
Update titles

rogeruiz committed:
Update draft for secrets landing

rogeruiz committed:
Draft progress for CF/Diego

rogeruiz committed:
Draft for rotating secrets passphrases

rogeruiz committed:
Clean up the steps a bit

rogeruiz committed:
Draft progress for rotating secrets;

Still needs a proper once-over for grammar and technical bits that
might be confusing to operators.

rogeruiz committed:
More progress on CF/Diego secrets

rogeruiz committed:
Initial progress on Bosh rotation

rogeruiz committed:
Fix CI build

rogeruiz committed:
Final draft for CF / Diego docs

rogeruiz committed:
More secrets updates for easier doing

cnelson commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,202 @@
+1. `uaa.clients.*.secrets` properties need to be coordinated across various Bosh
+   deployments and cloud.gov apps or service brokers.
+1. `collectd.riemann_server` property for Riemann server address.
+1. `nessus-agent.*` properties for Nessus.
```
This should be removed, riemann/collectd/nessus config is handled by runtime-config now

rogeruiz replied:
Ah perfect. thanks.

cnelson commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,202 @@
+   certificates and keys.
+    - The metadata comes from our external SAML providers. In order to rotate
+      these coordinate with them to exchange new metadata. The properties for
+      `login.saml.*` can be generate with the following documentation, although
```
login.saml cannot currently be rotated without breaking authentication with properly-configured IDPs.

We are waiting on https://www.pivotaltracker.com/n/projects/997278/stories/142082611 to complete to allow rotation without downtime.  Once that stories done, rotating this will be a multi-step process similar to the consul certs.

cnelson commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,202 @@
+      this data will also need to be coordinated and exchanged with all
+      providers.
+
+#### Credentials for RDS instances
```
Shouldn't this just get merged in from terraform and not be copy/pasted into secrets?

rogeruiz replied:
true that.

cnelson commented on content/docs/ops/runbook/rotating-secrets.md:
```diff
@@ -0,0 +1,187 @@
+  - `cf service "${name_of_deployer}"`
+1. Open the Fugacious link and save the credentials locally in the `credentials.yml`
+  - `cf service ${name_of_deployer} | grep Dashboard | awk '{ print $2 }'`
+1. If the deployer user is used for more than a `cf push`, give it the proper
```
I don't think we should do this;  If a deployer account needs special perms, then we should use an account that's defined in the cf manifest with those perms so it's documented, rather than hand modifying brokered deployer accounts.

rogeruiz replied:
yeah good point. are we worried at all about have those users extending the CF deployment? I'm thinking since most of these deployers exist in the same `org` and `space` we would only need a handful of the ones we currently have deployed.

cnelson replied:
I'd rather have a long, repeatable deployment than a fast inconsistent one :)  Also we can set override: false on all our users once created so UAA won't spend time recreating them on every startup

https://github.com/cloudfoundry/uaa-release/blob/develop/jobs/uaa/spec#L461-L463

rogeruiz replied:
I'll add these bits. @jmcarp added a new deployer user to the cf manifest, so I'll update this with that new process.

cnelson reviewed (commented):
Dang, that's a lot of documentation @rogeruiz;  Well done 👏 

Just a few things I think are incorrect / should be handed differently, thoughts?

wjwoodson committed:
Query for bosh secrets dependencies in 18F cg repos

rogeruiz committed:
Work in progress updates

rogeruiz committed:
More progress on Concourse

rogeruiz committed:
Rough draft of Rotating BOSH

rogeruiz committed:
Final draft for Rotating BOSH

rogeruiz committed:
Headings are hard

rogeruiz committed:
Update the order for Removing Access

rogeruiz committed:
Update titles to match order

rogeruiz committed:
Progress on Concourse rotation

rogeruiz committed:
Final draft for Concourse

jmcarp commented on content/docs/ops/runbook/rotating-bosh.md:
```diff
@@ -0,0 +1,194 @@
+### Agent limitations when updating BOSH
+
+The following should be referenced throughout the various deployments. [According
+to the BOSH development team](https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1492784585438034),
```
WDYT about filing a ticket at github.com/cloudfoundry/bosh for tracking purposes? Not sure they're paying slack to keep messages for long.

rogeruiz replied:
https://github.com/cloudfoundry/bosh/issues/1140 Found an oldie related to changing nats passwords. Will link to that.

rogeruiz replied:
Fixed in d46aababa2ce033ee9f047dbf53fc5f5fb2b8abe

jmcarp commented on content/docs/ops/runbook/rotating-bosh.md:
```diff
@@ -0,0 +1,194 @@
+
+#### Finding deployments with BOSH dependencies
+
+Search GitHub for the `BOSH_TARGET` and `BOSH_ENVIRONMENT` with a `YAML`
```
Could we also look at `bosh-cli deployments`?

jmcarp commented on content/docs/ops/runbook/rotating-bosh.md:
```diff
@@ -0,0 +1,194 @@
+Besides certificates, all secrets in each BOSH deployment can be generated and
+replaced or can be updated from [rotated IAM users]({{< relref "docs/ops/runbook/rotating-iam-users.md" >}}).
+
+#### Validating BOSH deployments
```
Do we need this section? At the moment, deploying a bosh instance also updates its cloud config, and if the director isn't available, updating cloud config will fail.

rogeruiz replied:
I was thinking this would at least be good to run just ensure that nothing is broken with the pipeline. Running the `common-releases-*` shouldn't be that big of a deal? wdyt?

jmcarp replied:
Agreed that it's not a big deal, but what does it buy us? My point is that the deploy-*-bosh tasks already include a step that exercises the director by updating its cloud config.

jmcarp commented on content/docs/ops/runbook/rotating-bosh.md:
```diff
@@ -0,0 +1,194 @@
+
+In the [`cg-deploy-bosh`](https://github.com/18F/cg-deploy-bosh) repository, you
+will find a `generate-master-bosh-certs.sh`. Create a `./tmp` directory
+and run that script in there with the IP address for Master BOSH. The IP address
```
This address shouldn't be in the secrets anymore--now it comes from terraform.

rogeruiz replied:
this is fixed now in 8b4036b55e728e15dd82f7cca9263f9d69fc6eb7

jmcarp commented on content/docs/ops/runbook/rotating-bosh.md:
```diff
@@ -0,0 +1,194 @@
+
+In the [`cg-deploy-bosh`](https://github.com/18F/cg-deploy-bosh) repository, you
+will find a `generate-bosh-certs.sh`. Create a `./tmp` directory and run that
+script in there with the IP address for Tooling BOSH. The IP address can be
```
Same point as above--these addresses shouldn't be in secrets now.

rogeruiz replied:
this is fixed now in 8b4036b55e728e15dd82f7cca9263f9d69fc6eb7

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,206 @@
+
+## Dealing with multiple secret files
+
+The Cloud Foundry manifest is split into two parts, a `main.yml` and
```
s/cloud foundry manifest/cloud foundry secrets/ ?

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
````diff
@@ -0,0 +1,206 @@
+```sh
+# note the target name `fr` here is the output of `fly targets`
+$ fly -t fr get-pipeline --pipeline deploy-cf | \
+  spurce json | \
````
s/spurce/spruce

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,206 @@
+
+1. Rotate all UAA user passwords with passwords matching the UAA restrictions
+   set in [the spec](https://github.com/18F/uaa-customized-boshrelease/) for our
+   `uaa-customized-boshrelease`.
```
Up to you, but I don't know that listing these requirements here is useful, since they might change in the future. More to the point, uaa will reject bad passwords, so operators might not need a list of rules here.

rogeruiz replied:
yeah the idea here is that since it's in a deployment, you won't get the _reject bad credentials bit_ until you deploy the thing, and it's not the best feedback loop. ➿

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
````diff
@@ -0,0 +1,206 @@
+  mv all-cert-certs production-cf-diego-certs
+```
+
+This script will generate all 25 of the certificates for both Diego and
````
Let's not hard-code certificate counts--the number doesn't really matter and will probably change over time.

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,206 @@
+#### Updating certificates in secrets
+
+Replace all the certificates corresponding to the filename and property matching
+in the secrets file, e.g. `certs_dir/hm9000_server.crt` maps to
```
hm9000 is gone now!

rogeruiz replied:
😅

jmcarp commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,206 @@
+
+1. `uaa.clients.*.secrets` properties need to be coordinated across various Bosh
+   deployments and cloud.gov apps or service brokers.
+1. `collectd.riemann_server` property for Riemann server address.
```
This should get merged from terraform.

jmcarp commented on content/docs/ops/runbook/rotating-secrets.md:
```diff
@@ -0,0 +1,190 @@
+org and space in order to create the cloud.gov service deployer account
+instance.
+
+### Auditing all CF deployment users
```
Can we drop this now that we're adding the shared broker deployer, or is it still useful?

rogeruiz replied:
Actually, I think so... Just for the kinds of users that only require a `cf push` 👍 I included this below it.

> Make sure that cloud.gov Service Account users are only used for basic `cf push`
deployments. If you're deploying service brokers, read the [Rotate Secrets V CF
external secrets section]({{< relref "docs/ops/runbook/rotating-]rotating-cloudfoundry-diego.md" >}}).

jmcarp commented on content/docs/ops/runbook/rotating-secrets.md:
```diff
@@ -0,0 +1,190 @@
+1. Delete the `${name_of_deployer}_prev` service account deployer.
+   - `cf delete-service ${name_of_deployer}_prev -f`
+
+## Rotate secrets passphrases
```
IMO this *has* to include adding passphrases back to keys in `concourse-credentials`, otherwise we have too much random cleanup to do afterwards.

rogeruiz replied:
Addressed in 3e20ce203c1fc18c59a1d9340f0db199f1da3e98

jmcarp commented on content/docs/ops/secrets.md:
```diff
@@ -66,21 +66,40 @@ All UAA clients and users and associated credentials should be created via the C
-    INPUT_FILE=tmp/cf.yml.enc OUTPUT_FILE=tmp/cf.yml PASSPHRASE=... ./decrypt.sh
+    INPUT_FILE=tmp/cf.main.yml.enc
+    OUTPUT_FILE=tmp/cf.main.yml
+    PASSPHRASE=$(fly --target <target> get-pipeline --pipeline deploy-cf | spruce json | jq -r '.resources[] | select(.name | test("common-prod")) | .sources.secrets_passphrase')
```
Is this a good time to standardize on `common-production`?

rogeruiz replied:
_yes_ 😉

rogeruiz replied:
I think this would be a good thing to update after the fact. I only touched on this file a bit to address `main` / `external` files.

rogeruiz committed:
Address 3/4 of feedback

h/t @cnelson

rogeruiz committed:
Update Bosh docs

rogeruiz committed:
Some CF/Diego updates

rogeruiz committed:
Update docs around external cf stuffz

rogeruiz committed:
Make it easier to list what's actually deployed.

https://github.com/18f/cg-site/pull/879#discussion_r119960624

rogeruiz committed:
Update some CF external things;

https://github.com/18f/cg-site/pull/879#discussion_r118766439

rogeruiz committed:
Point to GitHub instead of Slack;

https://github.com/18f/cg-site/pull/879#discussion_r119960472

rogeruiz committed:
Clean up the passphrase bits;

Though I don't know if there's anything to make this better for now...

rogeruiz committed:
Address the concourse credentials in the room;

https://github.com/18f/cg-site/pull/879#discussion_r119975467

rogeruiz committed:
Don't output what's already on the end of a link;

https://github.com/18f/cg-site/pull/879#discussion_r119974726

rogeruiz committed:
Remove validating BOSH deployment;

https://github.com/18f/cg-site/pull/879#discussion_r119974596

jcscottiii commented on content/docs/ops/runbook/rotating-cloudfoundry-diego.md:
```diff
@@ -0,0 +1,183 @@
+
+You will need to coordinate these changes with secrets and passphrase changes
+for various other deployments. You can [reference the list of external
+clients](https://github.com/18F/cg-deploy-cf/blob/master/cf-secrets-example.external.yml#L60-L87)
```
I wouldn't hardcode the lines for this section of YML. maybe put comments in the YML for which subset of clients you are talking about. (unless you mean all of the clients there, then just reference items under the `uaa.clients` key)

rogeruiz committed:
Try not to reference exact lines;

https://github.com/18F/cg-site/pull/879#discussion_r120637999

rogeruiz committed:
Stop breaking the build :trollface:

wjwoodson committed:
Reorder content to match order under Rotate Access heading

wjwoodson committed:
Add link Managing Users guide for removing admins
